packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,725 @@
+  }
+
+  @override
+  Stream<String> get logLines => _logLinesController.stream;
+}
+
+class CustomDevicePortForwarder extends DevicePortForwarder {
+  CustomDevicePortForwarder({
+    @required String deviceName,
+    @required List<String> forwardPortCommand,
+    @required RegExp forwardPortSuccessRegex,
+    this.numTries/*?*/,
+    @required ProcessManager processManager,
+    @required Logger logger,
+    Map<String, String> additionalReplacementValues = const <String, String>{}
+  }) : _deviceName = deviceName,
+       _forwardPortCommand = forwardPortCommand,
+       _forwardPortSuccessRegex = forwardPortSuccessRegex,
+       _processManager = processManager,
+       _processUtils = ProcessUtils(
+         processManager: processManager,
+         logger: logger
+       ),
+       _additionalReplacementValues = additionalReplacementValues;
+
+  final String _deviceName;
+  final List<String> _forwardPortCommand;
+  final RegExp _forwardPortSuccessRegex;
+  final ProcessManager _processManager;
+  final ProcessUtils _processUtils;
+  final int numTries;
+  final Map<String, String> _additionalReplacementValues;
+  final List<ForwardedPort> _forwardedPorts = <ForwardedPort>[];
+
+  @override
+  Future<void> dispose() async {
+    // copy the list so we don't modify it concurrently
+    List<ForwardedPort>.of(_forwardedPorts).forEach(unforward);
+  }
+
+  Future<ForwardedPort> _tryForward(int devicePort, int hostPort) async {
+    final List<String> interpolated = interpolateCommand(
+      _forwardPortCommand,
+      <String, String>{
+        'devicePort': '$devicePort',
+        'hostPort': '$hostPort'
+      },
+      additionalReplacementValues: _additionalReplacementValues
+    );
+
+    // launch the forwarding command
+    final Process process = await _processUtils.start(interpolated);
+
+    final Completer<ForwardedPort> completer = Completer<ForwardedPort>();
+
+    // read the outputs of the process, if we find a line that matches
+    // the configs forwardPortSuccessRegex, we complete with a successfully
+    // forwarded port
+    // Note that if that regex never matches, this will potentially run forever
+    // and the forwarding will never complete
+    final CustomDeviceLogReader reader = CustomDeviceLogReader(_deviceName)..listenToProcessOutput(process);
+    final StreamSubscription<String> logLinesSubscription = reader.logLines.listen((String line) {
+      if (_forwardPortSuccessRegex.hasMatch(line) && !completer.isCompleted) {
+        completer.complete(
+          ForwardedPort.withContext(hostPort, devicePort, process)
+        );
+      }
+    });
+
+    // if the process exists (even with exitCode == 0), that is considered
+    // a port forwarding failure and we complete with a null value.
+    unawaited(process.exitCode.then((int exitCode) {
+      if (!completer.isCompleted) {
+        completer.complete(null);
+      }
+    }));
+
+    // just some cleanup once the forwarding completed
```

Comment:
nit: this comment isn't that useful. Avoid comments that just describe exactly what the code below is doing.

Reply:
fixed it